tools/validate_whitespace.py

```diff
@@ -0,0 +1,100 @@
+#!/usr/bin/env python3
+import os
+import re
+import sys
+import glob
+import subprocess
+
+
+def load_ignores():
+    ignores = set()
+    fail = False
+    for ind, line in enumerate(open(".wsignore")):
+        line = line.strip()
+        if not line:
+            continue
+        if line.startswith("#"):
+            continue
+        expanded = glob.glob(line, recursive=True)
+        if len(expanded) == 0:
+            print(f".wsignore:{ind + 1}: No matches for line: {line}")
+            fail = True
+        ignores.update(expanded)
+    if fail:
+        sys.exit(1)
+    return ignores
+
+
+def count_trailing_newlines(s):
+    match = re.search(r"(\n+)$", s)
+    return len(match.group(1)) if match else 0
+
+
+def validate_contents(data):
+    if not data:
+        return
+    try:
+        text = data.decode("utf")
+    except Exception as ex:
+        yield f" Failed to decode utf-8: {ex}"
+        return
+
+    count = 0
+    max_trailing_errors = 9
+    for i, line in enumerate(text.split("\n")):
+        if not line:
+            continue
+        if line.strip() == "":
+            count += 1
+            if count <= max_trailing_errors:
+                yield f"{i + 1}: Whitespace-only line"
+            continue
+        if line.rstrip() != line:
+            count += 1
+            if count <= max_trailing_errors:
+                yield f"{i + 1}: Trailing whitespace {line[-200:]!r}"
+
+    if count > max_trailing_errors:
+        yield f" {count} cases of trailing whitespace"
+
+    newlines = count_trailing_newlines(text)
+
+    if newlines == 0:
+        yield " File does not end with a newline"
+
+    if newlines >= 2:
+        yield f" {newlines} newlines at the end"
+
+
+def main():
+    quiet = "-q" in sys.argv
+    files = subprocess.check_output(["git", "ls-files"], encoding="utf-8").split()
```

Comment:
You can run `git ls-files --exclude-from=` to let Git take care of exclusion.

Reply:
Does not work, it seems to be applied to untracked files only:

```
       -x <pattern>, --exclude=<pattern>
           Skip untracked files matching pattern. Note that pattern is a shell
           wildcard pattern. See EXCLUDE PATTERNS below for more information.

       -X <file>, --exclude-from=<file>
           Read exclude patterns from <file>; 1 per line.
```